x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtSignatureValidatorTests.java

```diff
@@ -0,0 +1,267 @@
+        Mockito.doAnswer(invocation -> {
+            safeAwait(reloadBarrier); // block here to ensure both threads have failed once
+
+            // grab the listener that is passed to reload method and call the onResponse method to simulate a successful reload
+            ActionListener<Tuple<Boolean, JwkSetLoader.JwksAlgs>> listener = invocation.getArgument(0);
+            if (reloadAttemptCounter.getAndIncrement() == 0) {
+                // first reload returns it was indeed reloaded
+                listener.onResponse(new Tuple<>(Boolean.TRUE, new JwkSetLoader.JwksAlgs(Collections.emptyList(), Collections.emptyList())));
+            } else {
+                // subsequent reloads returns it was not reloaded
+                listener.onResponse(
+                    new Tuple<>(Boolean.FALSE, new JwkSetLoader.JwksAlgs(Collections.emptyList(), Collections.emptyList()))
+                );
+            }
+            return null;
+        }).when(jwkSetLoader).reload(any(ActionListener.class));
+
+        Mockito.doAnswer(invocation -> {
+            // don't actually attempt to do the signature
+            // throw an exception to represent a failure or return null to represent a success
+            if (validateSignatureAttemptCounter.getAndIncrement() <= 1) {
+                // signature validation fails on first and second attempt, since we block when reloading we can be assured that first and
+                // second attempts are attempt1 for thread1 and thread2
+                throw new RuntimeException("boom");
+            } else {
+                // subsequent signature validation works
+                return null;
+            }
+        }).when(signatureValidator).validateSignature(any(SignedJWT.class), anyList());
+
+        final CyclicBarrier barrier = new CyclicBarrier(3);
+
+        Thread t1 = new Thread(() -> {
+            safeAwait(barrier);
+            signatureValidator.validate(randomIdentifier(), signedJWT, primaryListener);
+        });
+
+        Thread t2 = new Thread(() -> {
+            safeAwait(barrier);
+            signatureValidator.validate(randomIdentifier(), signedJWT, primaryListener);
+        });
+
+        t1.start();
+        t2.start();
+        safeAwait(barrier); // kick off the work
+        t1.join();
+        t2.join();
+
+        /**
+         * THESE VALUES REPRESENT HOW IT ACTUALLY WORKS, NOT HOW IT SHOULD WORK
+         * DUE THE BUG THE SECOND VALIDATION INCORRECTLY FAILS
+         */
+        assertThat(validateSignatureAttemptCounter.get(), is(3)); // should be 4, but the bug results in 1
+        assertThat(finalSuccessCounter.get(), is(1)); // should be 2, but the bug results in 1
+        assertThat(finalFailureCounter.get(), is(1)); // should be 0, but the bug results in 1
+        assertThat(reloadAttemptCounter.get(), is(2)); // 2 is correct with or without the bug
+    }
+
+    public void testJwtSignVerifyPassedForAllSupportedAlgorithms() {
```

Comment:
this was moved as-is from [JWTValidateUtilTests](https://github.com/elastic/elasticsearch/blob/v8.9.0/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtilTests.java#L31)